### PR TITLE
Implement PPO training

### PR DESCRIPTION
**Pull request by liujch1998**
Let's review and merge this one. I rebased my PPO implementations to the latest tip of master.

Ellen has graciously agreed to review the core PPO algo part, ETA Min 03/11.

Note: There are a few things that I intentionally did not implement, compared to TRL, because they are not needed:
- LR scheduler, gradient clipping (these are already baked in EasyLM's optimizer)
- in-batch random permutation of examples
- adaptive KL controller
- early_stopping (using fewer PPO ministeps per iteration if KL crosses threshold)
- ratio monitoring and skipping batch if ratio is larger than a fixed threshold

### CODE REVIEW

File: EasyLM/models/llama/llama_model.py [hunk 1]
```diff
@@ -918,14 +918,14 @@ def __call__(
...
+
         if position_ids is None:
             if past_key_values is not None:
                 raise ValueError("Make sure to provide `position_ids` when passing `past_key_values`.")
 
-            position_ids = jnp.broadcast_to(jnp.arange(sequence_length)[None, :], (batch_size, sequence_length))
-
-        if attention_mask is None:
-            attention_mask = jnp.ones((batch_size, sequence_length))
+            position_ids = jnp.broadcast_to(jnp.clip(jnp.cumsum(attention_mask, axis=-1) - 1, a_min=0), (batch_size, sequence_length))
```
Comment (liujch1998) [inline EasyLM/models/llama/llama_model.py:929]: The position_ids was not set correctly if the sequences are not left-aligned. Fixed it here.

File: EasyLM/data.py [hunk 2]
```diff
@@ -268,6 +280,66 @@ def vocab_size(self):
...
+        config.path = 'openbmb/UltraFeedback'
+        config.name = ''
+        config.split = 'train'
+        config.streaming = False
+        config.seq_length = 1024
+        config.num_workers = 8
+        config.batch_size = 8
+        config.policy_prefix_tokens = '<|user|>\n'
+        config.policy_suffix_tokens = '\n<|assistant|>\n'
+        config.reward_prefix_tokens = 'Human: '
```
Comment (hamishivi) [inline EasyLM/data.py]: The current RMs I train expect tulu format, maybe we should swap to using that as the default?

File: scripts/tpu_vm_setup.sh [hunk 3]
```diff
@@ -9,13 +9,15 @@ sudo apt-get update && sudo apt-get install -y \
     nodejs \
     bmon \
     p7zip-full \
-    nfs-common
+    nfs-common \
+    golang
```
Comment (hamishivi) [inline scripts/tpu_vm_setup.sh]: Where does the golang requirement come from?

Comment (liujch1998) [reply scripts/tpu_vm_setup.sh]: It is a dependency of jax-smi, which I'm using to trace the TPU usage. Does this look OK?

Comment (hamishivi) [reply scripts/tpu_vm_setup.sh]: no worries, just making sure a random dependency hadn't snuck in :) having jax-smi included by default is a good idea.

File: EasyLM/models/llama/llama_train_ppo.py [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+        input_ids=prompt_input_ids,
+        attention_mask=prompt_attn_mask,
+        generation_config=generation_config,
+        params=policy_train_state.params['params'],
+        prng_key=rng_generator(),
+    )
+    input_ids = outputs.sequences # (B, L)
+
+    # NOTE: This is a hack because generate() weirdly forces the last token to be 0 instead of 2
+    last_token_index = jnp.argmax(jnp.cumsum(jnp.where(input_ids == pad_token_id, 0, 1), axis=1), axis=1) # (B)
```
Comment (ellenmellon) [inline EasyLM/models/llama/llama_train_ppo.py:217]: Is this the same as `last_token_index = jnp.sum(jnp.where(input_ids == pad_token_id, 0, 1), axis=1)`?

File: EasyLM/models/llama/llama_train_ppo.py [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+            loss_fn = lambda policy_params, value_params: ppo_loss(
+                policy_model, value_model,
+                policy_params, value_params,
+                rng_generator(),
+                mb_input_ids, mb_attn_mask, mb_cont_input_ids, mb_cont_attn_mask, mb_cont_logps, mb_cont_values, mb_advantages, mb_returns,
+            )
+            grad_fn = jax.value_and_grad(loss_fn, argnums=[0, 1], has_aux=True)
+            (_, stats), (policy_grads, value_grads) = grad_fn(policy_train_state.params, value_train_state.params)
+            policy_train_state = policy_train_state.apply_gradients(grads=policy_grads)
+            value_train_state = value_train_state.apply_gradients(grads=value_grads)
```
Comment (ellenmellon) [inline EasyLM/models/llama/llama_train_ppo.py:311]: Do you change the policy/value parameters after each mini_batch? I think the reason why I used mini_batch_size for training and a bigger batch_size for getting rollouts in fgrlhf was to fill out all gpu usage bubbles. So ideally, the model parameters should only update after the actual/big batch_size. In pytorch, we did `.backward()` after each mini_batch to accumulate gradients, and `.step()` after each ppo_epoch.

Comment (liujch1998) [reply EasyLM/models/llama/llama_train_ppo.py:311]: Thanks Ellen! I referred to the TRL impl, which does gradient update for each minibatch: https://github.com/huggingface/trl/blob/v0.7.4/trl/trainer/ppo_trainer.py#L761-L800

My understanding was that batch_size is set to be larger than mini_batch_size to parallelize and speed up rollouts. Doing either way (apply gradient update per batch or per mini_batch) seems fine to me.

In addition, in my experiments I always set batch_size == minibatch_size :)
